src/twisted/python/failure.py

```diff
@@ -163,8 +160,8 @@ def __init__(self, frameinfo, back):
         name, filename, lineno, localz, globalz = frameinfo
         self.f_code = _Code(name, filename)
         self.f_lineno = lineno
-        self.f_globals = {}
-        self.f_locals = {}
+        self.f_globals = dict(globalz or {})
+        self.f_locals = dict(localz or {})
```

Comment:
I think this `or` is untested.  Is it meant to handle `None`?  What if one of these values is `{}`?  Substituting a new empty dict in place of the given empty dict is probably surprise.  Also, why substitute an empty dict at all?  If the frameinfo had None, shouldn't the `_Frame` have None?

Reply:
The old unit test used to pass None in. I suppose that means the or is not tested now.
That opens up the question of what should be passable as an argument.
I don't know whether the real frameinfo can have a None.

Certainly taking `None` or `{}` as an input and replacing them with a new empty dict maintains the existing behaviour
which always gave a new empty dict.
Also I think it makes sense to always produce a new dict, just to avoid mutability issues, hence the `dict()` call.

Reply:
Okay - "always a copy" is reasonable.